### PR TITLE
Solving Hi-res download blocks PNG

### PR DESCRIPTION
**Pull request by Ramanathi**
Regarding #2069
@ewpatton

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

File: appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public NewYoungAndroidProjectWizard(final ProjectToolbar toolbar) {
       @Override
       public boolean validate(String value) {
         errorMessage = TextValidators.getErrorMessage(value);
-        if (errorMessage.length()>0){
+        if (errorMessage.length()>0){
```
Comment (Ramanathi) [inline appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java:59]: Ignore this!

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 2]
```diff
@@ -304,6 +304,7 @@ function PNG() {
  * @const
  */
 PNG.HEADER = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A];
+var pHY_data =  [0, 0, 16, 25, 0, 0, 16, 25, 1];
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:307]: These values should be prefixed with `0x` as they are intended to be hex, not decimal.

File: appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java (hunk 1 above)
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java:59]: Not sure why this is changed in this commit.

File: appinventor/appengine/src/com/google/appinventor/client/youngandroid/TextValidators.java [hunk 3]
```diff
@@ -77,6 +77,7 @@ public static boolean checkNewProjectName(String projectName) {
   public static boolean checkNewComponentName(String componentName) {
 
     // Check that it meets the formatting requirements.
+
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/youngandroid/TextValidators.java:80]: Revert the changes in this file. They are not relevant to what you're trying to accomplish in this PR.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 4]
```diff
@@ -376,12 +377,14 @@ PNG.prototype.processData_ = function(data) {
...
   }
   this.chunks = [];
+  var phy = [112,72,89,115];
   while (chunkStart < data.length) {
     var length = decode4();
     var type = read4();
     var chunkData = readData(length);
     chunkStart += length;
     var crc = decode4();
+    if (type == 'IDAT') {this.chunks.push(new PNG.Chunk(9, 'pHYs', pHY_data, crc32(phy.concat(pHY_data)))); }
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:387]: Code style: each statement should be on its own line.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 5]
```diff
@@ -376,12 +377,14 @@ PNG.prototype.processData_ = function(data) {
     return data.slice(chunkStart, chunkStart + length);
   }
   this.chunks = [];
+  var phy = [112,72,89,115];
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:380]: Code style: spaces after commas.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 6]
```diff
@@ -461,11 +464,11 @@ Blockly.exportBlockAsPng = function(block) {
     img.src = uri;
     img.onload = function() {
       var canvas = document.createElement('canvas');
-      canvas.width = img.width;
-      canvas.height = img.height;
+      canvas.width = 2*(img.width);
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:467]: Parentheses are not necessary here. Code style: space around operators.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 7]
```diff
@@ -461,11 +464,11 @@ Blockly.exportBlockAsPng = function(block) {
...
+      canvas.height = 2*(img.height);
       var context = canvas.getContext('2d');
-      context.drawImage(img, 0, 0);
-
+      context.drawImage(img, 0, 0,img.width,img.height,0,0,canvas.width,canvas.height);
+
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:471]: Leave blank lines blank, no whitespace.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 8]
```diff
@@ -461,11 +464,11 @@ Blockly.exportBlockAsPng = function(block) {
...
     img.onload = function() {
       var canvas = document.createElement('canvas');
-      canvas.width = img.width;
-      canvas.height = img.height;
+      canvas.width = 2*(img.width);
+      canvas.height = 2*(img.height);
       var context = canvas.getContext('2d');
-      context.drawImage(img, 0, 0);
-
+      context.drawImage(img, 0, 0,img.width,img.height,0,0,canvas.width,canvas.height);
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:470]: Code style: spaces after commas.

File: appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java (hunk 1 above)
Comment (Ramanathi) [reply appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java:59]: Please Ignore this, I forgot to form a branch When solving this issue.

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 4 above)
Comment (Ramanathi) [reply appinventor/blocklyeditor/src/exportBlocksImage.js:387]: Got it!

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 2 above)
Comment (Ramanathi) [reply appinventor/blocklyeditor/src/exportBlocksImage.js]: Fixed

File: appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java (hunk 1 above)
Comment (Ramanathi) [reply appinventor/appengine/src/com/google/appinventor/client/wizards/youngandroid/NewYoungAndroidProjectWizard.java]: Fixed

File: appinventor/appengine/src/com/google/appinventor/client/youngandroid/TextValidators.java (hunk 3 above)
Comment (Ramanathi) [reply appinventor/appengine/src/com/google/appinventor/client/youngandroid/TextValidators.java]: Fixed

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 5 above)
Comment (Ramanathi) [reply appinventor/blocklyeditor/src/exportBlocksImage.js]: Done

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 6 above)
Comment (Ramanathi) [reply appinventor/blocklyeditor/src/exportBlocksImage.js]: Fixed

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 8 above)
Comment (Ramanathi) [reply appinventor/blocklyeditor/src/exportBlocksImage.js]: Done

**Comment by Ramanathi** [comment]
Hello @ewpatton I've updated the PR!

**Comment by ewpatton** [comment]
@Ramanathi Thanks, however on my mac I'm getting an error saying there isn't enough image data. I'm not sure why that is, as if I remove the phys chunk it does the right thing. I'll need to look into the spec to understand why it doesn't think there's enough data.

**Comment by ewpatton** [comment]
So diving in a bit more deeply, I've realized the issue arises from how you insert the pHYs chunk into the PNG. Mainly, there can be more than one IDAT chunk, and the pHYs chunk must come before the first IDAT. In my sample images I was generating, there was more than one IDAT chunk, so your code inserted more than one pHYs chunk. This patch made it work correctly on my machine:

```diff
diff --git a/appinventor/blocklyeditor/src/exportBlocksImage.js b/appinventor/blocklyeditor/src/exportBlocksImage.js
index d17780974d..0ebf836eaa 100644
--- a/appinventor/blocklyeditor/src/exportBlocksImage.js
+++ b/appinventor/blocklyeditor/src/exportBlocksImage.js
@@ -377,16 +377,12 @@ PNG.prototype.processData_ = function(data) {
     return data.slice(chunkStart, chunkStart + length);
   }
   this.chunks = [];
-  var phy = [112, 72, 89, 115];
   while (chunkStart < data.length) {
     var length = decode4();
     var type = read4();
     var chunkData = readData(length);
     chunkStart += length;
     var crc = decode4();
-    if (type == 'IDAT') {
-      this.chunks.push(new PNG.Chunk(9, 'pHYs', pHY_data, crc32(phy.concat(pHY_data))));
-    }
     this.chunks.push(new PNG.Chunk(length, type, chunkData, crc));
   }
 };
@@ -473,6 +469,14 @@ Blockly.exportBlockAsPng = function(block) {

       function download(png) {
         png.setCodeChunk(code);
+        for (var i = 0; i < png.chunks.length; i++) {
+          if (png.chunks[i].type == 'IDAT') {
+            var phy = [112, 72, 89, 115];
+            png.chunks.splice(i, 0, new PNG.Chunk(9, 'pHYs', pHY_data,
+              crc32(phy.concat(pHY_data))));
+            break;
+          }
+        }
         var blob = png.toBlob();
         var a = document.createElement('a');
         a.download = (block.getChildren().length === 0 ? block.type : 'blocks') + '.png';
```

**Comment by Ramanathi** [comment]
> So diving in a bit more deeply, I've realized the issue arises from how you insert the pHYs chunk into the PNG. Mainly, there can be more than one IDAT chunk, and the pHYs chunk must come before the first IDAT. In my sample images I was generating, there was more than one IDAT chunk, so your code inserted more than one pHYs chunk.

@ewpatton If that's the case, I've Introduced flag to rectify this

**Comment by ewpatton** [comment]
@AppInventorWorkerBee ok to test

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 9]
```diff
@@ -376,12 +377,18 @@ PNG.prototype.processData_ = function(data) {
     return data.slice(chunkStart, chunkStart + length);
   }
   this.chunks = [];
+  var phy = [112, 72, 89, 115];
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:380]: I still don't like doing this in `processData_` because it is used on multiple code paths, not just in the PNG generation. Instead, I'd prefer if we move the logic of adding the chunk to the `download` function below, similar to how `setCodeChunk` is handled. Lastly, we should also check to see if there is a phys chunk and replace it. At the moment, browsers don't appear to be adding the chunk, but in the off chance they do so in the future we want to replace the existing chunk rather than have 2 in conflict.

File: appinventor/blocklyeditor/src/exportBlocksImage.js (hunk 7 above)
Comment (ewpatton) [reply appinventor/blocklyeditor/src/exportBlocksImage.js:475]: Still need this whitespace cleaned up.

**Comment by Ramanathi** [comment]
@ewpatton Done!

**Comment by Ramanathi** [comment]
@ewpatton I've updated the Code Style.
Is there any reason that this type of Code style to be followed?

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 10]
```diff
@@ -461,13 +462,23 @@ Blockly.exportBlockAsPng = function(block) {
...
-      context.drawImage(img, 0, 0);
+      context.drawImage(img, 0, 0, img.width, img.height, 0, 0, canvas.width, canvas.height);
 
       function download(png) {
         png.setCodeChunk(code);
+        for (var i = 0; i < png.chunks.length; i++) {
+          var phy = [112, 72, 89, 115];
+          if (png.chunks[i].type == 'pHYs') {
+            png.chunks.splice(i, 1, new PNG.Chunk(9, 'pHYs', pHY_data, crc32(phy.concat(pHY_data)))); //replacing existing pHYs chunk
+        	break;
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:476]: Not sure why the break is indented here. It should be at the same indentation as the previous line.

File: appinventor/blocklyeditor/src/exportBlocksImage.js [hunk 11]
```diff
@@ -461,13 +462,23 @@ Blockly.exportBlockAsPng = function(block) {
...
+          } else if (png.chunks[i].type == 'IDAT') {
+            png.chunks.splice(i, 0, new PNG.Chunk(9, 'pHYs', pHY_data, crc32(phy.concat(pHY_data)))); // adding new pHYs chunk
+        	break;
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/exportBlocksImage.js:479]: Ditto.

**Comment by Ramanathi** [comment]
> Is there any reason that this type of Code style to be followed?
@ewpatton ?

**Comment by ewpatton** [comment]
> Is there any reason that this type of Code style to be followed?

@Ramanathi We follow the Google Java styleguide for any Java code (exception: user-facing names in components have the first letter capitalized) and the Google JavaScript style guide for JavaScript code. The recommendations I made were in line with the JavaScript style guide.
